File: 백준/Silver/10974. 모든 순열/모든 순열.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "ex.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const n = Number(input[0]);

let visited = new Array(n).fill(false);
let selected = new Array(n).fill(0);
let answer = "";

function dfs(depth) {
  if (depth === n) {
    let result = [];
    for (let i = 0; i < n; i++) {
      result.push(selected[i]);
    }
    answer += result.join(" ") + "\n";
  }
  for (let i = 1; i <= n; i++) {
    if (visited[i]) continue;

    selected[depth] = i;
    visited[i] = true;
    dfs(depth + 1);
    visited[i] = false;
  }
}

dfs(0);

console.log(answer);

File: 백준/Silver/10974. 모든 순열/모든 순열.py
n = int(input())

ans = []
visited = [0]*(n+1)
def dfs(depth):
    if depth == n:
        print(' '.join(map(str, ans)))
        return

    for i in range(1, n+1):
        if visited[i]:
            continue
        visited[i] = 1
        ans.append(i)
        dfs(depth+1)
        ans.pop()
        visited[i] = 0

dfs(0)